workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/WorkflowLifecycleOwner.kt

```diff
@@ -0,0 +1,220 @@
+     * 2. Someone eventually calls [destroyOnDetach], which will either schedule the lifecycle to
+     *    destroyed if the view is attached, or destroy it immediately if it's detached.
+     *
+     * If this is not done, any observers registered with the [Lifecycle] may be leaked as they will
+     * never see the destroy event.
+     *
+     * @param findParentLifecycle A function that is called whenever [view] is attached, and should
+     * return the [Lifecycle] to use as the parent lifecycle. If not specified, defaults to looking
+     * up the view tree by calling [ViewTreeLifecycleOwner.get] on [view]'s parent, and if none is
+     * found, then looking up [view]'s context wrapper chain for something that implements
+     * [LifecycleOwner]. This only needs to be passed if [view] will be used as the root of a new
+     * view hierarchy, e.g. for a new dialog. If no parent lifecycle is found, then the lifecycle
+     * will become [RESUMED] when it's attached for the first time, and stay in that state until
+     * it is re-attached with a non-null parent or [destroyOnDetach] is called.
+     */
+    public fun installOn(
+      view: View,
+      findParentLifecycle: () -> Lifecycle? = { findViewTreeLifecycle(view) }
+    ) {
+      RealWorkflowLifecycleOwner(view, findParentLifecycle).also {
+        ViewTreeLifecycleOwner.set(view, it)
+        view.addOnAttachStateChangeListener(it)
+      }
+    }
+
+    /**
+     * Looks for the nearest [ViewTreeLifecycleOwner] on [view] and returns it if it's an instance
+     * of [WorkflowLifecycleOwner]. Convenience function for retrieving the owner set by
+     * [installOn].
+     */
+    public fun get(view: View): WorkflowLifecycleOwner? =
+      ViewTreeLifecycleOwner.get(view) as? WorkflowLifecycleOwner
+
+    private fun findViewTreeLifecycle(view: View): Lifecycle? {
+      // Start at our view's parent – if we look on our own view, we'll just get this back.
+      return (view.parent as? View)?.let(::lifecycleOwnerFromViewTreeOrContext)?.lifecycle
+    }
+  }
+}
+
+/**
+ * @param enforceMainThread Allows disabling the main thread check for testing.
+ */
+@OptIn(WorkflowUiExperimentalApi::class)
+@VisibleForTesting(otherwise = PRIVATE)
+internal class RealWorkflowLifecycleOwner(
+  private val view: View,
+  private val findParentLifecycle: () -> Lifecycle?,
+  enforceMainThread: Boolean = true,
+) : WorkflowLifecycleOwner,
+  LifecycleOwner,
+  OnAttachStateChangeListener,
+  LifecycleEventObserver {
+
+  private val localLifecycle =
+    if (enforceMainThread) LifecycleRegistry(this) else createUnsafe(this)
+
+  /**
+   * The parent lifecycle found by calling [ViewTreeLifecycleOwner.get] on the owning view's parent
+   * (once it's attached), or if no [ViewTreeLifecycleOwner] is set, then by trying to find a
+   * [LifecycleOwner] on the view's context.
+   *
+   * When the view is detached, we keep the reference to, and observing, the previous parent
```

Comment:
```suggestion
   * When the view is detached, we keep the reference to, and observe, the previous parent
```

Reply:
The parent was already observed. I wanted this to say "we keep observing the previous parent."